addons/l10n_latam_document/models/account_invoice.py

```diff
@@ -0,0 +1,430 @@
+                self, journal, invoice_type, partner):
+            return super(
+                AccountInvoice, self)._get_available_document_types(
+                    journal, invoice_type, partner)
+        """
+        for invoice in self:
+            res = invoice._get_available_document_types(
+                invoice.journal_id, invoice.type, invoice.partner_id)
+            invoice.l10n_latam_available_document_type_ids = res[
+                'available_document_types']
+
+    @api.multi
+    def write(self, vals):
+        """
+        If someone change the type (for eg from sale order), we update
+        de document type
+        """
+        inv_type = vals.get('type')
+        # if len(vals) == 1 and vals.get('type'):
+        # podrian pasarse otras cosas ademas del type
+        if inv_type:
+            for rec in self:
+                res = rec._get_available_document_types(
+                    rec.journal_id, inv_type, rec.partner_id)
+                vals['l10n_latam_document_type_id'] = res[
+                    'document_type'].id
+                # call write for each inoice
+                super(AccountInvoice, rec).write(vals)
+                return True
+        return super(AccountInvoice, self).write(vals)
+
+    @api.model
+    def _get_available_document_types(
+            self, journal, invoice_type, partner):
+        """ This function is to be inherited by different localizations and
+        MUST return a dictionary with two keys:
+        * 'available_document_types': available document types on
+        this invoice
+        * 'document_type': suggested document type on this invoice
+        """
+        if not journal.l10n_latam_use_documents:
+            return {
+                'available_document_types':
+                    self.env['l10n_latam.document.type'],
+                'document_type':
+                    self.env['l10n_latam.document.type'],
+            }
+        raise UserError(_(
+            'Method not implemented by localization of %s') % (
+                journal.company_id.country_id.name))
+
+    @api.constrains('l10n_latam_document_type_id', 'l10n_latam_document_number')
+    @api.onchange('l10n_latam_document_type_id', 'l10n_latam_document_number')
+    def validate_document_number(self):
+        # if we have a sequence, number is set by sequence and we dont
+        # check this
+        for rec in self.filtered(
+                lambda x: not x.l10n_latam_sequence_id and x.l10n_latam_document_type_id):
+            l10n_latam_document_number = \
+                rec.l10n_latam_document_type_id._format_document_number(
+                    rec.l10n_latam_document_number)
+            if l10n_latam_document_number != rec.l10n_latam_document_number:
+                rec.l10n_latam_document_number = l10n_latam_document_number
+
+    @api.constrains('type', 'l10n_latam_document_type_id')
+    def check_invoice_type_document_type(self):
+        for rec in self:
```

Comment:
```suggestion
        for rec in self.filtered('l10n_latam_document_type_id.internal_type'):
```

To avoid the first `if`